includes/libraries/parsecsv.lib.php

```diff
@@ -0,0 +1,1163 @@
+                $sort_type = SORT_NUMERIC;
+            } elseif ($this->sort_type == 'string') {
+                $sort_type = SORT_STRING;
+            }
+
+            ($this->sort_reverse) ? krsort($rows, $sort_type) : ksort($rows, $sort_type);
+
+            if ($this->offset !== null || $this->limit !== null) {
+                $rows = array_slice($rows, ($this->offset === null ? 0 : $this->offset), $this->limit, true);
+            }
+        }
+
+        if (!$this->keep_file_data) {
+            $this->file_data = null;
+        }
+
+        return $rows;
+    }
+
+    /**
+     * Create CSV data from array
+     *
+     * @access public
+     * @param   data        2D array with data
+     * @param   fields      field names
+     * @param   append      if true, field names will not be output
+     * @param   is_php      if a php die() call should be put on the first
+     *                      line of the file, this is later ignored when read.
+     * @param   delimiter   field delimiter to use
+     *
+     * @return  CSV data (text string)
+     */
+    public function unparse($data = array(), $fields = array(), $append = false, $is_php = false, $delimiter = null) {
+        if (!is_array($data) || empty($data)) {
+            $data = &$this->data;
+        }
+
+        if (!is_array($fields) || empty($fields)) {
+            $fields = &$this->titles;
+        }
+
+        if ($delimiter === null) {
+            $delimiter = $this->delimiter;
+        }
+
+        $string = ($is_php) ? "<?php header('Status: 403'); die(' '); ?>" . $this->linefeed : '';
+        $entry = array();
+
+        // create heading
+        if ($this->heading && !$append && !empty($fields)) {
+            foreach ($fields as $key => $value) {
+                $entry[] = $this->_enclose_value($value, $delimiter);
+            }
+
+            $string .= implode($delimiter, $entry) . $this->linefeed;
+            $entry = array();
+        }
+
+        // create data
+        foreach ($data as $key => $row) {
+            foreach ($row as $field => $value) {
+                $entry[] = $this->_enclose_value($value, $delimiter);
+            }
+
+            $string .= implode($delimiter, $entry) . $this->linefeed;
+            $entry = array();
+        }
+
+        if ($this->convert_encoding) {
+            $string = iconv($this->input_encoding, $this->output_encoding, $string);
+        }
+
+        return $string;
+    }
+
+    /**
+     * Load local file or string
+     *
+     * @access public
+     * @param   input   local CSV file
+     *
+     * @return  true or false
+     */
+    public function load_data($input = null) {
+        $data = null;
+        $file = null;
+
+        if (is_null($input)) {
+            $file = $this->file;
+        } elseif (file_exists($input)) {
+            $file = $input;
+        } else {
+            $data = $input;
+        }
+
+        if (!empty($data) || $data = $this->_rfile($file)) {
```

Comment:
That library is abandoned, and hasn't been updated in a year. If we're going to use this library we're going to have to assume responsibility for fixing the bugs in it because that library is never going to get updated again. The author hasn't logged into GitHub on his personal account or any organization he has access to in 8 months.